vhost_user_backend, vm-virtio, vmm: Close leaked file descriptors from 'epoll::create()'

likebreath opened the pull request:
This patch fixes file descriptor leak related to epoll::create() from
various virtio devices.

Fixes: #1124

Signed-off-by: Bo Chen <chen.bo@intel.com>

sboeuf commented on vm-virtio/src/block.rs:
```diff
@@ -767,6 +767,7 @@ impl<T: DiskFile> BlockEpollHandler<T> {
             }
         }
 
+        epoll::close(epoll_fd).map_err(DeviceError::EpollClose)?;
```
Problem with this approach is that you're gonna close the file descriptor only if things are Ok(()). You want to make sure that even if this function returns earlier than expected (because of an Err()), the file descriptor will be closed.

One simple way to do it would be to get a `File` from the `epoll_fd` `RawFd`. The created `File` would be dropped at the end of the function and `close()` will be automatically called on this `File`.

There's also another way, and we should consider this for the long term plan. There's a new crate on `rust-vmm` called `event_manager`, which abstract a bunch of the complexity and duplication related to epoll events. We want to switch to this eventually, and I think this crate won't leak any file descriptor based on the way it's implemented.
https://github.com/rust-vmm/event-manager

sboeuf replied:
/cc @rbradford @sameo

sboeuf replied:
@likebreath please go ahead with the quick solution of using `let epoll_file = unsafe{ File::from_raw_fd(epoll_fd) };`. Every place where the code was expecting `epoll_fd`, you can use `epoll_file.as_raw_fd()`. This way we make sure at the end of the scope, the `File` will take care of closing the file descriptor.
I opened an issue https://github.com/cloud-hypervisor/cloud-hypervisor/issues/1183 to track the move to `event-manager` crate from rust-vmm.

likebreath commented:
PTAL. I just adjusted PR based on comments. Also, I added another commit to fix the fd leak from `vmm` and `vhost_user_backend`. Thanks! @sboeuf @rbradford

sboeuf commented on vm-virtio/src/vsock/unix/muxer.rs:
```diff
@@ -793,6 +793,13 @@ impl VsockMuxer {
+        // Ignore the result because there is nothing we can do about it.
+        let _ = epoll::close(self.epoll_fd);
+    }
+}
```
I would change the epoll_fd stored in the VsockMuxer struct into an epoll_file. And everytime the RawFd is needed, just call into self.epoll_file.as_raw_fd().
This way, no need to implement the Drop trait, as the file descriptor will be closed when the VsockMuxer instance is destroyed.

likebreath replied:
Thanks for the detailed comments. The PR is updated and rebased. Any more questions or comments, please let me know.

sboeuf commented on vhost_user_backend/src/lib.rs:
```diff
@@ -424,6 +424,13 @@ impl VringWorker {
+        // Ignore the result because there is nothing we can do about it.
+        let _ = epoll::close(self.epoll_fd);
+    }
+}
```
Same here, store a File in the structure instead of the RawFd directly. This way, no need for a special Drop handling.

sboeuf commented on vmm/src/lib.rs:
```diff
@@ -174,6 +174,13 @@ impl AsRawFd for EpollContext {
+        // Ignore the result because there is nothing we can do about it.
+        let _ = epoll::close(self.raw_fd);
+    }
+}
```
Ditto.

likebreath commented:
The `test_serial_off` test is failing, and I am looking at it.

likebreath commented:
@sboeuf Thanks for fixing the test failure on `test_serial_off`, and merging the PR. :)
